src/signal.c

```diff
@@ -15,113 +16,171 @@
 
 #include <libswiftnav/signal.h>
 
-typedef struct {
-  u8 constellation;
-  u8 band;
-  u16 sat_count;
-  u16 (*sat_get)(u32 local_index);
-  u32 (*local_index_get)(u16 sat);
-} sat_group_t;
-
-static const char * constellation_strs[CONSTELLATION_COUNT] = {
-  [CONSTELLATION_GPS] = "GPS",
-  [CONSTELLATION_SBAS] = "SBAS"
+static const u16 number_of_sats[] = {
+    NUM_SATS_GPS,
+    NUM_SATS_SBAS,
 };
 
-static const char * band_strs[BAND_COUNT] = {
-  [BAND_L1] = "L1"
+static const u16 first_sat_index[] = {
+    0,              /* GPS  */
+    NUM_CODES_GPS,  /* SBAS */
 };
 
-static const char * unknown_str = "?";
+static const u16 first_sat_id[] = {
+    GPS_FIRST_PRN,     /* GPS  */
+    SBAS_FIRST_PRN,    /* SBAS */
+};
 
-static u16 sat_get_gps(u32 local_index);
-static u16 sat_get_sbas(u32 local_index);
-static u32 local_index_get_gps(u16 sat);
-static u32 local_index_get_sbas(u16 sat);
-
-static const sat_group_t sat_groups[CONSTELLATION_COUNT] = {
-  [CONSTELLATION_GPS] = {
-    CONSTELLATION_GPS, BAND_L1, NUM_SATS_GPS,
-     sat_get_gps, local_index_get_gps
-  },
-  [CONSTELLATION_SBAS] = {
-    CONSTELLATION_SBAS, BAND_L1, NUM_SATS_SBAS,
-    sat_get_sbas, local_index_get_sbas
-  }
+static const u8 codes_on_cons[] = {
+    CODES_ON_GPS_SAT,  /* GPS  */
+    CODES_ON_SBAS_SAT, /* SBAS */
 };
 
-static u16 sat_get_gps(u32 local_index)
-{
-  return local_index + GPS_FIRST_PRN;
-}
+static const u8 codes_on_gps[CODES_ON_GPS_SAT] = {
+    CODE_GPS_L1CA,
+    CODE_GPS_L2CM,
+};
 
-static u16 sat_get_sbas(u32 local_index)
-{
-  return local_index + SBAS_FIRST_PRN;
-}
+static const u8 codes_on_sbas[CODES_ON_SBAS_SAT] = {
+    CODE_SBAS_L1CA,
+};
 
-static u32 local_index_get_gps(u16 sat)
-{
-  return sat - GPS_FIRST_PRN;
-}
+static const char * unknown_str = "?";
 
-static u32 local_index_get_sbas(u16 sat)
+gnss_signal_t construct_sid(enum code code, u16 sat)
 {
-  return sat - SBAS_FIRST_PRN;
+  gnss_signal_t sid = { .code = code, .sat = sat};
+  return sid;
 }
 
 int sid_to_string(char *s, int n, gnss_signal_t sid)
 {
-  const char *constellation_str = (sid.constellation < CONSTELLATION_COUNT) ?
-      constellation_strs[sid.constellation] : unknown_str;
-  const char *band_str = (sid.band < BAND_COUNT) ?
-      band_strs[sid.band] : unknown_str;
+  const char *code_str = unknown_str;
```

Comment:
At that point we had different approach for the code enum. I'll change this back to table of strings.